### ISSUE TITLE
group_writer.cpp:623: [realm-core-6.0.8] Assertion failed: free_ref < locked.ref || free_ref >= (locked.ref + locked.size) with (locked.ref, locked.size, locked.released_at_version, free_ref, m_current_version, m_alloc.get_file_path_for_assertions()) = [1

### ISSUE THREAD

**Issue by destryteeter** [realm/realm-cocoa#6636]
## Goals
Resolve crash issue

## Expected Results
No crash

## Actual Results
Crashed:
```
/Users/realm/workspace/realm_realm-core_release_6.0.8/src/realm/group_writer.cpp:623: [realm-core-6.0.8] Assertion failed: free_ref < locked.ref || free_ref >= (locked.ref + locked.size) with (locked.ref, locked.size, locked.released_at_version, free_ref, m_current_version, m_alloc.get_file_path_for_assertions()) =  [137359000, 1960, 8649, 137359000, 8650, "/private/var/mobile/Containers/Shared/AppGroup/F286316C-BF09-41E5-9F14-F518772C50EE/default.realm"]
0   People Power Family                 0x000000010639e850 _ZN5realm4utilL18terminate_internalERNSt3__118basic_stringstreamIcNS1_11char_traitsIcEENS1_9allocatorIcEEEE + 28
1   People Power Family                 0x000000010639ecec _ZN5realm4util19terminate_with_infoEPKcS2_lS2_OSt16initializer_listINS0_9PrintableEE + 388
2   People Power Family                 0x00000001062e8dfc _ZN5realm4util19terminate_with_infoIJRKmRKyRmS6_RyNSt3__112basic_stringIcNS8_11char_traitsIcEENS8_9allocatorIcEEEEEEEvPKciSG_SG_DpOT_ + 0
3   People Power Family                 0x00000001062e890c _ZN5realm11GroupWriter17recreate_freelistEm + 2420
4   People Power Family                 0x00000001062e71bc _ZN5realm11GroupWriter11write_groupEv + 776
5   People Power Family                 0x00000001062df4c4 _ZN5realm2DB16low_level_commitEyRNS_11TransactionE + 380
6   People Power Family                 0x00000001062df2d8 _ZN5realm2DB9do_commitERNS_11TransactionE + 112
7   People Power Family                 0x00000001062df8a8 _ZN5realm11Transaction27commit_and_continue_as_readEv + 72
8   People Power Family                 0x0000000105ced81c _ZN5realm5_impl16RealmCoordinator12commit_writeERNS_5RealmE + 276
9   People Power Family                 0x0000000106002490 _ZN5realm5Realm18commit_transactionEv + 440
10  People Power Family                 0x0000000105f0ffb4 -[RLMRealm commitWriteTransactionWithoutNotifying:error:] + 604
11  People Power Family                 0x0000000105f0fd48 -[RLMRealm commitWriteTransaction:] + 88
12  People Power Family                 0x0000000105761944 $sSo24PPProfessionalMonitoringC15Presence_FamilyE17refreshCallCenteryyys5Error_pSgcFZySo06PPCallG0CSg_AFtcfU_ + 13684
13  People Power Family                 0x00000001056c4e4c $sSo12PPCallCenterCSgs5Error_pSgIeggg_ACSo7NSErrorCSgIeyByy_TR + 152
14  People Power Family                 0x0000000105afec74 __58+[PPProfessionalMonitoring getCallCenter:userId:callback:]_block_invoke_3 + 76
15  libdispatch.dylib                   0x000000010e79e338 _dispatch_call_block_and_release + 24
16  libdispatch.dylib                   0x000000010e79f730 _dispatch_client_callout + 16
17  libdispatch.dylib                   0x000000010e7ad710 _dispatch_main_queue_callback_4CF + 976
18  CoreFoundation                      0x00000001a1d6f7fc 50CF3336-313F-3A7D-9048-CB1ED8EC3368 + 690172
19  CoreFoundation                      0x00000001a1d6a6d0 50CF3336-313F-3A7D-9048-CB1ED8EC3368 + 669392
20  CoreFoundation                      0x00000001a1d69ce8 CFRunLoopRunSpecific + 424
21  GraphicsServices                    0x00000001abeb438c GSEventRunModal + 160
22  UIKitCore                           0x00000001a5e98444 UIApplicationMain + 1932
23  People Power Family                 0x0000000105072ac0 main + 148
24  libdyld.dylib                       0x00000001a1bf18f0 3D6D64B4-CB2B-3CC4-A7E9-02774DF7AE74 + 6384!!! IMPORTANT: Please send this log and info about Realm SDK version and other relevant reproduction info to help@realm.io.2020-07-07 17:25:25.831999-0700 People Power Family[849:177872] /Users/realm/workspace/realm_realm-core_release_6.0.8/src/realm/group_writer.cpp:623: [realm-core-6.0.8] Assertion failed: free_ref < locked.ref || free_ref >= (locked.ref + locked.size) with (locked.ref, locked.size, locked.released_at_version, free_ref, m_current_version, m_alloc.get_file_path_for_assertions()) =  [137359000, 1960, 8649, 137359000, 8650, "/private/var/mobile/Containers/Shared/AppGroup/F286316C-BF09-41E5-9F14-F518772C50EE/default.realm"]
```

## Steps to Reproduce
Launched app, run for an extended period of time making various calls to write/read the realm database (30-60 minutes).

## Code Sample
Source code is private and cannot be shared.

## Version of Realm and Tooling
Please update to the latest Carthage version: 0.35.0. You currently are on 0.34.0
```
ProductName:	Mac OS X
ProductVersion:	10.15.5
BuildVersion:	19F101

/Applications/Xcode.app/Contents/Developer
Xcode 11.5
Build version 11E608c

/usr/local/bin/pod
1.9.1
Realm (5.2.0)

/bin/bash
GNU bash, version 3.2.57(1)-release (x86_64-apple-darwin19)

/usr/local/bin/carthage
0.34.0
(not in use here)

/usr/bin/git
git version 2.24.3 (Apple Git-128)
```

**Comment by tgoyne** [comment]
@finnschiermer This is an assertion failure that I haven't seen before.

**Comment by cferg-entale** [comment]
Did you discover the cause of this?

**Comment by cferg-entale** [comment]
I am also experiencing a crash in my code after an extended period of time (30+ mins). My app is an audio app and write playback progress to Realm every 2 seconds. Any recommendations are very welcome.
```
/Users/realm/workspace/realm_realm-core_release_6.0.9/src/realm/array_basic_tpl.hpp:123: [realm-core-6.0.9] Assertion failed: ndx < m_size [5, 0]
0   Realm                               0x0000000109cf8474 _ZN5realm4utilL18terminate_internalERNSt3__118basic_stringstreamIcNS1_11char_traitsIcEENS1_9allocatorIcEEEE + 28
1   Realm                               0x0000000109cf8718 _ZN5realm4util9terminateEPKcS2_lOSt16initializer_listINS0_9PrintableEE + 328
2   Realm                               0x000000010976db80 _ZN5realm4util9terminateIJmmEEEvPKcS3_lDpT_ + 144
3   Realm                               0x0000000109c80e98 _ZN5realm3Obj3setIfEERS0_NS_6ColKeyET_b + 1052
4   Realm                               0x000000010986a584 _ZN12_GLOBAL__N_18setValueIfEEvP13RLMObjectBaseN5realm6ColKeyET_ + 84
5   Realm                               0x000000010986a524 _ZZZN12_GLOBAL__N_110makeSetterIffEEP11objc_objectP11RLMPropertyEUb0_ENKUlvE_clEv + 120
6   Realm                               0x000000010986a4a0 ___ZN12_GLOBAL__N_110makeSetterIffEEP11objc_objectP11RLMProperty_block_invoke_2 + 240
7   Entale                              0x0000000104d02314 $s6Entale15PlaybackHistoryC18updateProgressData33_64BC4DC07B202437F4FE0DBD10E3DC61LL9contentId8position8durationyAA07ContentP0O_S2dSgtFy10RealmSwift0T0VXEfU0_ + 204
8   Entale                              0x0000000104cea6cc $s10RealmSwift0A0V6EntaleE5writeyyyACKXEKFyyXEfU_ + 96
9   Entale                              0x0000000104ce9ddc $s10RealmSwift0A0V6EntaleE9safeWrite33_9A8FC24798E030663DEFD3E32840878DLLyyyyKXEKF + 148
10  Entale                              0x0000000104cea570 $s10RealmSwift0A0V6EntaleE5writeyyyACKXEKF + 140
11  Entale                              0x0000000104f568c8 $s6Entale19DefaultRealmServiceC5writeyyy0C5Swift0C0VKXEKF + 128
12  Entale                              0x0000000104f57ec8 $s6Entale19DefaultRealmServiceCAA0cD0A2aDP5writeyyy0C5Swift0C0VKXEKFTW + 24
13  Entale                              0x0000000104d01e94 $s6Entale15PlaybackHistoryC18updateProgressData33_64BC4DC07B202437F4FE0DBD10E3DC61LL9contentId8position8durationyAA07ContentP0O_S2dSgtF + 1136
14  Entale                              0x0000000104d019f8 $s6Entale15PlaybackHistoryC15progressUpdated33_64BC4DC07B202437F4FE0DBD10E3DC61LLyyAA16DataNotificationVyAA0B6UpdateVGFyycfU_ + 432
15  Entale                              0x0000000104cfdce0 $sIeg_IeyB_TR + 56
16  libdispatch.dylib                   0x000000010b2097fc _dispatch_call_block_and_release + 24
17  libdispatch.dylib                   0x000000010b20abd8 _dispatch_client_callout + 16
18  libdispatch.dylib                   0x000000010b218c34 _dispatch_main_queue_callback_4CF + 1316
19  CoreFoundation                      0x000000019994c5e4 7519E999-1053-3367-B9D5-8844F6D3BDC6 + 693732
20  CoreFoundation                      0x00000001999475d8 7519E999-1053-3367-B9D5-8844F6D3BDC6 + 673240
21  CoreFoundation                      0x0000000199946adc CFRunLoopRunSpecific + 464
22  GraphicsServices                    0x00000001a38e7328 GSEventRunModal + 104
23  UIKitCore                           0x000000019da5463c UIApplicationMain + 1936
24  Entale                              0x0000000104cd72fc main + 128
25  libdyld.dylib                       0x00000001997d0360 7B531A15-3E73-3185-90E2-B88D9476DA5E + 4960!!! IMPORTANT: Please send this log and info about Realm SDK version and other relevant reproduction info to help@realm.io.2020-07-16 17:57:55.755841+0100 Entale[809:304792] /Users/realm/workspace/realm_realm-core_release_6.0.9/src/realm/array_basic_tpl.hpp:123: [realm-core-6.0.9] Assertion failed: ndx < m_size [5, 0]
0   Realm                               0x0000000109cf8474 _ZN5realm4utilL18terminate_internalERNSt3__118basic_stringstreamIcNS1_11char_traitsIcEENS1_9allocatorIcEEEE + 28
1   Realm                               0x0000000109cf8718 _ZN5realm4util9terminateEPKcS2_lOSt16initializer_listINS0_9PrintableEE + 328
2   Realm                               0x000000010976db80 _ZN5realm4util9terminateIJmmEEEvPKcS3_lDpT_ + 144
3   Realm                               0x0000000109c80e98 _ZN5realm3Obj3setIfEERS0_NS_6ColKeyET_b + 1052
4   Realm                               0x000000010986a584 _ZN12_GLOBAL__N_18setValueIfEEvP13RLMObjectBaseN5realm6ColKeyET_ + 84
5   Realm                               0x000000010986a524 _ZZZN12_GLOBAL__N_110makeSetterIffEEP11objc_objectP11RLMPropertyEUb0_ENKUlvE_clEv + 120
6   Realm                               0x000000010986a4a0 ___ZN12_GLOBAL__N_110makeSetterIffEEP11objc_objectP11RLMProperty_block_invoke_2 + 240
7   Entale                              0x0000000104d02314 $s6Entale15PlaybackHistoryC18updateProgressData33_64BC4DC07B202437F4FE0DBD10E3DC61LL9contentId8position8durationyAA07ContentP0O_S2dSgtFy10RealmSwift0T0VXEfU0_ + 204
8   Entale                              0x0000000104cea6cc $s10RealmSwift0A0V6EntaleE5writeyyyACKXEKFyyXEfU_ + 96
9   Entale                              0x0000000104ce9ddc $s10RealmSwift0A0V6EntaleE9safeWrite33_9A8FC24798E030663DEFD3E32840878DLLyyyyKXEKF + 148
10  Entale                              0x0000000104cea570 $s10RealmSwift0A0V6EntaleE5writeyyyACKXEKF + 140
11  Entale                              0x0000000104f568c8 $s6Entale19DefaultRealmServiceC5writeyyy0C5Swift0C0VKXEKF + 128
12  Entale                              0x0000000104f57ec8 $s6Entale19DefaultRealmServiceCAA0cD0A2aDP5writeyyy0C5Swift0C0VKXEKFTW + 24
13  Entale                              0x0000000104d01e94 $s6Entale15PlaybackHistoryC18updateProgressData33_64BC4DC07B202437F4FE0DBD10E3DC61LL9contentId8position8durationyAA07ContentP0O_S2dSgtF + 1136
14  Entale                              0x0000000104d019f8 $s6Entale15PlaybackHistoryC15progressUpdated33_64BC4DC07B202437F4FE0DBD10E3DC61LLyyAA16DataNotificationVyAA0B6UpdateVGFyycfU_ + 432
15  Entale                              0x0000000104cfdce0 $sIeg_IeyB_TR + 56
16  libdispatch.dylib                   0x000000010b2097fc _dispatch_call_block_and_release + 24
17  libdispatch.dylib                   0x000000010b20abd8 _dispatch_client_callout + 16
18  libdispatch.dylib                   0x000000010b218c34 _dispatch_main_queue_callback_4CF + 1316
19  CoreFoundation                      0x000000019994c5e4 7519E999-1053-3367-B9D5-8844F6D3BDC6 + 693732
20  CoreFoundation                      0x00000001999475d8 7519E999-1053-3367-B9D5-8844F6D3BDC6 + 673240
21  CoreFoundation                      0x0000000199946adc CFRunLoopRunSpecific + 464
22  GraphicsServices                    0x00000001a38e7328 GSEventRunModal + 104
23  UIKitCore                           0x000000019da5463c UIApplicationMain + 1936
24  Entale                              0x0000000104cd72fc main + 128
25  libdyld.dylib                       0x00000001997d0360 7B531A15-3E73-3185-90E2-B88D9476DA5E + 4960!!! IMPORTANT: Please send this log and info about Realm SDK version and other relevant reproduction info to help@realm.io.
(lldb) 
```

**Comment by destryteeter** [comment]
Right, similar behavior where we pull data from our cloud every 10-60 seconds and update our local realm database.

Caught another crash while running the app from Xcode:
```
realm_coordinator.cpp

void RealmCoordinator::commit_write(Realm& realm)
{
    REALM_ASSERT(!m_config.immutable());
    REALM_ASSERT(realm.is_in_transaction());

    Transaction& tr = Realm::Internal::get_transaction(realm);
    {
        // Need to acquire this lock before committing or another process could
        // perform a write and notify us before we get the chance to set the
        // skip version
        util::CheckedLockGuard l(m_notifier_mutex);

        tr.commit_and_continue_as_read(); >> Thread 1: EXC_BAD_ACCESS (code=1, address=0x2d7e00004)
...

--- thread ---
#9	0x0000000104af17e8 in closure #1 in static PPLocationUser.refreshUsers(_:) at /Users/destry/Developer/PPC/iOS/presence-family-ios/Presence Family/Models/PFLocationUser.swift:229
#8	0x00000001052fe310 in -[RLMRealm commitWriteTransaction:] at /Users/destry/Developer/PPC/iOS/presence-family-ios/Pods/Realm/Realm/RLMRealm.mm:686
#7	0x00000001052fe57c in -[RLMRealm commitWriteTransactionWithoutNotifying:error:] at /Users/destry/Developer/PPC/iOS/presence-family-ios/Pods/Realm/Realm/RLMRealm.mm:698
#6	0x00000001053f0a58 in realm::Realm::commit_transaction() at /Users/destry/Developer/PPC/iOS/presence-family-ios/Pods/Realm/Realm/ObjectStore/src/shared_realm.cpp:671
#5	0x00000001050dbde4 in realm::_impl::RealmCoordinator::commit_write(realm::Realm&) at /Users/destry/Developer/PPC/iOS/presence-family-ios/Pods/Realm/Realm/ObjectStore/src/impl/realm_coordinator.cpp:688
#4	0x00000001056cde70 in realm::Transaction::commit_and_continue_as_read() ()
#3	0x00000001056cd8a0 in realm::DB::do_commit(realm::Transaction&) ()
#2	0x00000001056cda70 in realm::DB::low_level_commit(unsigned long long, realm::Transaction&) ()
#1	0x00000001056d479c in realm::GroupWriter::GroupWriter(realm::Group&, realm::DBOptions::Durability) ()
#0	0x000000010562114c in realm::Array::init_from_mem(realm::MemRef) ()


```

**Comment by destryteeter** [comment]
It looks like the same crash as far as I can see.  The only difference is the initial report happened on a App Store release build.  While the more recent one happened when running the app locally.

**Comment by bmunkholm** [comment]
The first one asserts, but not the 2'nd and not in the same function, right?
But they surely could be related. Let's just keep it here for now.

Could you possibly setup some stress testing of your app?
If you get to a point where you can reproduce, would it be possible to share it with us privately?

**Comment by cferg-entale** [comment]
@bmunkholm, thank you for your comment. I have created an issues [here](https://github.com/realm/realm-cocoa/issues/6656).

**Comment by RealmBot** [comment]
➤ Finn Andersen commented:

If there are sightings of this bug in scenarios without realm-sync, please speak up.

**Comment by destryteeter** [comment]
We're not using realm-sync.

I should note that the crash does not reproduce with realm 3.2.  Only after updating the realm dependency does the crash start appearing.

**Comment by finnschiermer** [comment]
Thx for info re realm-sync, that allows us to rule out a bunch of code :-)

Did you mean version 5.3.2 ?

**Comment by finnschiermer** [comment]
Likely duplicate of https://github.com/realm/realm-core/issues/3434, just looks different because the underlying code has changed.

**Comment by finnschiermer** [comment]
Goes back 2 years, but is still here despite all that's changed. Is not related to Sync.

**Comment by destryteeter** [comment]
I was using 5.2.0.  Did see the same crasher on newer releases though.

**Comment by pavel-ship-it** [comment]
Hi @destryteeter ,
Is it possible to reproduce that issue on the latest release?

**Comment by finnschiermer** [comment]
Possibly fixed by https://github.com/realm/realm-core/issues/3904

**Comment by geragray** [comment]
@destryteeter could you please try reproducing the issue with the latest Realm Swift 5.4.0 (in includes fix from https://github.com/realm/realm-core/issues/3904)

**Comment by destryteeter** [comment]
Checking.  Running the app for about 1.5 hours now and so far so good.

**Comment by destryteeter** [comment]
~4 hours in.  Xcode crashed on me (Some apple bug I think) so I will start my endurance test over.

**Comment by destryteeter** [comment]
Alright, I think you got it.  Running Realm 5.4.0 for 10+ hours without a crash.

Thank you for this fix.  This will help with my transition from using Cocoapods to using Carthage (which doesn't support Realm 3.x) to manage my dependencies and greatly increase my build times against the Realm framework.

**Comment by bmunkholm** [comment]
Awesome @destryteeter! I'll close this one then. Let us know if you run into anything else or have any other improvement suggestions.
